### PR TITLE
docs(autocomplete): Adds option group example

### LINKED ISSUE #10196: Example about mat-optgroup inside mat-autocomplete
#### Bug, feature request, or proposal:

Documentation request

#### What is the expected behavior?

Having an working example of this code :
 [https://material.angular.io/components/autocomplete/overview#option-groups](https://material.angular.io/components/autocomplete/overview#option-groups)

#### What is the current behavior?

No stackblitz or demo. And it's not easy to understand... because for the "mat-optgroup" you need (implicitly ?) to go to the "Select" component section, and with a group structure (eg: pokemonGroups) the autocomplete filter (eg: filteredOptions: Observable<string[]>;) doesn't work anymore.

#### What are the steps to reproduce?

https://stackblitz.com/edit/angular-nwdj9u

#### What is the use-case or motivation for changing an existing behavior?

Having a autocomplete input with group.

Like this for example 

Input : "d" <---------- autocomplete 
[**D** <------------------- group
     Dog <------------- option1
     Dragon <---------- option2
     ...
]

### PR DESCRIPTION
**Pull request by raygervais**
fixes #10196, 
StackBlitz example: https://stackblitz.com/edit/angular-smg2xm?file=app/app.component.ts

Adds a option group example which displays ideal logic and best practices for the material autocomplete component when using said groups.

### CODE REVIEW

File: src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import { Component, OnInit } from '@angular/core';
+import { FormGroup, FormBuilder, Validators } from '@angular/forms';
+import { Observable } from 'rxjs/Observable';
+import { startWith } from 'rxjs/operators/startWith';
```
Comment (crisbeto) [inline src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts]: In general we don't do spaces around the imports. E.g. these should be `import {startWith} from 'rxjs/operators/startWith';`

File: src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  
+  constructor(private fb: FormBuilder) {
+    this.stateForm = this.fb.group({
+      stateGroup: [''],
+    });
+   }
+  
+  ngOnInit() {
+    this.stateGroupOptions = this.stateForm.get('stateGroup').valueChanges
+    .pipe(
```
Comment (crisbeto) [inline src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts]: The `.pipe` should either be on the previous line or be indented.

File: src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    names: ['Vermont', 'Virginia']
+  }, {
+    letter: 'W',
+    names: ['Washingto n', 'West Virginia', 'Wisconsin', 'Wyoming']
+  }];
+
+  stateGroupOptions: Observable<StateGroup[]>;
+  
+  constructor(private fb: FormBuilder) {
+    this.stateForm = this.fb.group({
```
Comment (crisbeto) [inline src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts]: You can initialize this one when you declare the property.

File: src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    names: ['Tennessee', 'Texas']
+  }, {
+    letter: 'U',
+    names: ['Utah']
+  }, {
+    letter: 'V',
+    names: ['Vermont', 'Virginia']
+  }, {
+    letter: 'W',
+    names: ['Washingto n', 'West Virginia', 'Wisconsin', 'Wyoming']
```
Comment (crisbeto) [inline src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts]: Extra space before the "n".

**Comment by raygervais** [comment]
Just noticed the lint and e2e failures, will address today.

**Comment by raygervais** [comment]
Sorry for the delay, I've been wrestling school obligations and apartment hunting prior to starting a new job. Does anyone have any idea how to fix the lint error here? I have not seen it before: 

`ERROR: /home/travis/build/angular/material2/src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts[2, 33]: 'Validators' is declared but its value is never read.`

Here is the code that I'm trying to use: 
```ts
stateForm: FormGroup = this.fb.group({
  stateGroup: '',
});
```

**Comment by rafaelss95** [comment]
Just remove `Validators` from `import`.

File: src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.html [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<form [formGroup]="stateForm" novalidate>
```
Comment (rafaelss95) [inline src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.html]: `novalidate` isn't necessary.

File: src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    return this.stateGroups;
+  }
+
+  private _filter(opt: string[], val: string) {
+    const filterValue = val.toLowerCase();
+    return opt.filter(item => item.toLowerCase().startsWith(filterValue));
+  }
+}
+
+class StateGroup {
```
Comment (rafaelss95) [inline src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts]: You could replace this by an `Interface` and move this to the top of the file (after imports);

**Comment by raygervais** [comment]
Thanks for the tips @rafaelss95, and sorry for the delay. I've fixed the Lint / E2E issues, and the failing test case relates to `Material Table`. If there is any other fix needed, I'd be happy to correct 👍

**Pull request merged**